Code/GraphMol/ChemTransforms/ChemTransforms.cpp

```diff
@@ -319,24 +331,49 @@ ROMol *replaceCore(const ROMol &mol, const ROMol &coreQuery,
 
   unsigned int origNumAtoms = mol.getNumAtoms();
   std::vector<int> matchingIndices(origNumAtoms, -1);
+  std::vector<int> allIndices(origNumAtoms, -1);
+  //  std::cerr << "replaceDummies " << (int) replaceDummies << std::endl;
   for (MatchVectType::const_iterator mvit = matchV.begin();
        mvit != matchV.end(); mvit++) {
     if (replaceDummies ||
         coreQuery.getAtomWithIdx(mvit->first)->getAtomicNum() > 0) {
+      //      std::cerr << "matchingIndices[" << mvit->second << "] = " << mvit->first << std::endl;
       matchingIndices[mvit->second] = mvit->first;
     }
+    allIndices[mvit->second] = mvit->first;
   }
 
   RWMol *newMol = static_cast<RWMol *>(new ROMol(mol));
   std::vector<Atom *> keepList;
   std::map<int, Atom *> dummyAtomMap;
   unsigned int nDummies = 0;
+
+  // go through the matches in query order, not target molecule
+  //  order
+  std::vector<std::pair<int,int> > matchorder_atomidx;
   for (unsigned int i = 0; i < origNumAtoms; ++i) {
+    int queryatom = allIndices[i];
+    matchorder_atomidx.push_back( std::make_pair(queryatom, i) );
+  }
+
+  std::sort(matchorder_atomidx.begin(), matchorder_atomidx.end());
```

Comment:
Yes, the default comparator for pairs does the right thing  sort by first break ties by second.